CMSIS-NN wrapper update for interface changes

felix-johnny opened the pull request:
Background: CMSIS-NN int8  APIs are changed
where the pass by value function arguments
are replaced by pass by reference struct
arguments.

The following changes are done in TFL micro

1. Update int8 cmsis-nn's depthwise conv,
   fully connected and average pooling
   wrapper files to use the new  interface.

2. For the above mentioned operators,
   updates from the reference implementation
   (e.g, CalculateOpData() in Prepare() instead
   of Eval()) are ported as well.

felix-johnny committed:
CMSIS-NN wrapper update for interface changes

Background: CMSIS-NN int8  APIs are changed
where the pass by value function arguments
are replaced by pass by reference struct
arguments.

The following changes are done in TFL micro

1. Update int8 cmsis-nn's depthwise conv,
   fully connected and average pooling
   wrapper files to use the new  interface.

2. For the above mentioned operators,
   updates from the reference implementation
   (e.g, CalculateOpData() in Prepare() instead
   of Eval()) are ported as well.

felix-johnny commented:
sorry about the delete-restore cycle there! Re-based and adapted to new change in cmsis.inc where new build files have to be explicitly added.

felix-johnny commented:
@njeffrie Hi Nat! Hope all is well. Could you have a look at this PR? thanks in advance :-)

felix-johnny commented:
Rebased to resolve merge conflicts.

felix-johnny commented:
@njeffrie Hi Nat! Could you have a look at the changes ?

njeffrie commented on tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:
```diff
@@ -156,12 +93,62 @@ TfLiteStatus AverageEvalInt8(TfLiteContext* context, const TfLiteNode* node,
+    filter_dims.w = params->filter_width;
+    filter_dims.c = 1;
+
+    auto* buffer_idx = reinterpret_cast<int*>(node->user_data);
```
Prefer using OpData properly here, adding buffer_index to OpData and using data->buffer_index like in fully_connected.cc

felix-johnny replied:
Sure. As part of the change, I moved the CalcOpData() and checks to the prepare phase similar to fully_connected and conv's as well.

njeffrie replied:
Sounds good.  I don't see the changes here yet - I assume this should become OpData* data = node->user_data, then using data->buffer_idx.

njeffrie commented on tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:
```diff
@@ -387,7 +352,7 @@ TfLiteRegistration* Register_AVERAGE_POOL_2D() {
 TfLiteRegistration* Register_MAX_POOL_2D() {
   static TfLiteRegistration r = {/*init=*/pooling::Init,
                                  /*free=*/nullptr,
-                                 /*prepare=*/pooling::Prepare,
```
Why set Prepare to nullptr?  Aren't you doing valuable checks?

felix-johnny replied:
well, the prepare() was just allocating a scratch buffer irrespective of the pooling type. And max pool doesn't require additional buffers.

I've split the Prepare as AveragePrepare() and MaxPrepare() , with checks , CalcOpData() and scratch buffer allocation only for the right pooling type.

njeffrie replied:
I see, this looks much better, thanks.

felix-johnny committed:
Max pool & Review comment update.

1. Review comments are taken in
2. Interface for max pool is updated.

felix-johnny commented on tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:
```diff
@@ -173,89 +182,112 @@ void MaxEvalFloat(TfLiteContext* context, TfLiteNode* node,
                          GetTensorData<uint8_t>(output));
 }
 
 TfLiteStatus MaxEvalInt8(TfLiteContext* context, const TfLiteNode* node,
```
I have updated the max pool to the new interface as well(The CMSIS part wasn't ready when I first made this PR).

njeffrie commented on tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:
```diff
@@ -186,153 +182,150 @@ void MaxEvalFloat(TfLiteContext* context, TfLiteNode* node,
+  filter_dims.w = params->filter_width;
+  filter_dims.c = 1;
 
   auto* buffer_idx = reinterpret_cast<int*>(node->user_data);
```
Can we switch to using node->user_data->buffer_idx here as well?

felix-johnny replied:
Ah..my bad..I had missed updating the usage of buffer_idx here. This would have been an issue in a full network. I've updated it now. Thanks!

felix-johnny committed:
Review comment fix: Incorrect assignment of  buffer_idx

buffer_idx was incorrectly initialized in the Eval routines
of Average and Max pool. This is fixed.

felix-johnny commented:
Thanks for the reviews Nat @njeffrie !
